src/js/components/tasks/retrieveDataset/data.tsx

```diff
@@ -0,0 +1,518 @@
+
+  const clearSearch = () => {
+    setSearchValue('');
+    setSearch('');
+  };
+
+  const searchLabel = t('Search for objects or fields');
+
+  return (
+    <Search
+      className="slds-text-body_regular slds-m-bottom_small"
+      assistiveText={{ label: searchLabel }}
+      placeholder={searchLabel}
+      inlineHelpText={truncatedMsg}
+      value={searchValue}
+      clearable={Boolean(searchValue)}
+      onChange={handleSearch}
+      onClear={clearSearch}
+    />
+  );
+};
+
+export const SchemaList = ({
+  type,
+  heading,
+  schema,
+  checkedChanges,
+  errors,
+  handleSelectGroup,
+  handleChange,
+  className,
+  ...props
+}: {
+  type: 'all' | 'selected';
+  heading: string;
+  schema: DatasetSchema;
+  checkedChanges?: Changeset;
+  errors?: string;
+  handleSelectGroup?: (thisGroup: Changeset, checked: boolean) => void;
+  handleChange?: ({
+    groupName,
+    change,
+    checked,
+  }: {
+    groupName: string;
+    change: string;
+    checked: boolean;
+  }) => void;
+  className?: string;
+}) => {
+  const { t } = useTranslation();
+  const [expandedPanels, setExpandedPanels] = useState<BooleanObject>({});
+  const [filteredSchema, setFilteredSchema] = useState(schema);
+  const [search, setSearch] = useState<string>('');
+
+  let truncatedMsg = null;
+  let sortedSchema = sortSchema(filteredSchema);
+  const noSchema = isEmpty(filteredSchema);
+  if (sortedSchema.length > SCHEMA_SIZE_LIMIT) {
+    sortedSchema = sortedSchema.slice(0, SCHEMA_SIZE_LIMIT);
+    truncatedMsg = t(
+      'Only displaying the first {{limit}} objects. Enter a term above to search the full list.',
+      { limit: SCHEMA_SIZE_LIMIT },
+    );
+  }
+
+  const handlePanelToggle = (groupName: string) => {
+    setExpandedPanels({
+      ...expandedPanels,
+      [groupName]: !expandedPanels[groupName],
+    });
+  };
+
+  // Update schema when search term changes
+  useEffect(() => {
+    setFilteredSchema(
+      search ? filterSchema(schema, search.toLowerCase()) : schema,
+    );
+  }, [schema, search]);
+
+  let emptyHeading =
+    type === 'all' ? t('No data available to retrieve') : t('No data selected');
+  let emptyMsg =
+    type === 'all'
+      ? null
+      : t(
+          'Choose objects or fields from the left panel to add them to this dataset.',
+        );
+  if (search) {
+    emptyHeading = t('No data found');
+    emptyMsg = t(
+      'Change or remove your search term above to view additional data.',
+    );
+  }
+
+  const paddingClasses =
+    'slds-m-left_xx-small slds-p-horizontal_medium slds-p-right_medium';
+
+  return (
+    <div
+      className={classNames('has-checkboxes', 'slds-form', className)}
+      data-form="task-retrieve-changes"
+      {...props}
+    >
+      <ModalCard
+        heading={
+          <div className="slds-m-left_xx-small">
+            <SearchForm truncatedMsg={truncatedMsg} setSearch={setSearch} />
+            <p>{heading}</p>
+          </div>
+        }
+        noBodyPadding
+      >
```

Comment:
@jgerigmeyer this may be something we've done before, but this is the first time I noticed that we aren't supposed to have anything but inline elements in the ModalCard heading: https://react.lightningdesignsystem.com/components/cards/#prop-heading

Reply:
👀 @stacyk I think the only place we're breaking this currently is the "read-only" badge when a user without push permissions is an Epic collaborator (it contains a single `div`):

![Screen Shot 2022-07-25 at 12 09 10 PM](https://user-images.githubusercontent.com/552316/180824699-bacf93f6-a72a-4c7f-be0c-29537a33a63b.jpg)

I'm not sure how big of a deal this is?

In this new case, we could pretty easily move the search input just below the heading instead of inside/above it. What do you think?

Reply:
@jgerigmeyer Re: read only badge: this is something we could note in handoff, but I don't think we should update it at this time.

As far as the other one, I think we could do the same (as in, note it in some handoff documentation)

Reply:
@stacyk Do you think it's worth moving the search input below the heading in this case? I could do that if you prefer.

Reply:
@stacyk @SondraE Here's what it looks like with search moved below the header:

![Screen Shot 2022-07-26 at 10 55 25 AM](https://user-images.githubusercontent.com/552316/181039619-cabe5e0b-3a42-4916-9ba7-dfddb286ab5e.jpg)